common/flatpak-prune.c

```diff
@@ -0,0 +1,698 @@
+
+#include <glib/gi18n-lib.h>
+
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <stdio.h>
+#include <errno.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/file.h>
+
+#include "flatpak-error.h"
+#include "flatpak-prune-private.h"
+#include "flatpak-variant-impl-private.h"
+#include "libglnx/libglnx.h"
+#include "valgrind-private.h"
+
+/* This is a custom implementation of ostree-prune that caches the
+ * traversal for better performance on larger repos. It also merges the list-object
+ * and prune operation to avoid allocating a lot of memory for the list of all
+ * objects in the repo.
+ *
+ * Locking strategy:
+ *
+ * Ostree supports three kinds of approaches to handling parallel access to
+ * the repo.
+ *
+ * EXCLUSIVE LOCK:
+ *  All global operations that modify the repo state take an exclusive lock on the
+ *  repo which means no other repo-modifying operation is allowed in parallel. This
+ *  is currently only done for pruneing and summary generation. Prune for instance
+ *  is global it traverses from a set of root commits and assumes that everything
+ *  that isn't reachable can be deleted, which is not compatible with adding a
+ *  new commit that doesn't have a root commit yet.
+ *  NOTE: Whenever objects are deleted we always hold an exclusive lock.
+ *
+ * SHARED LOCKS:
+ *  Operations that do local modifications take a shared lock. This means we can
+ *  have multiple such operations in parallel with each other, but not in parallel
+ *  with an exclusive lock. The typical operation that does this is the commit.
+ *  During a commit we don't add to the transaction objects that already exist
+ *  in the repo, so we rely on them not disappearing because then when we finally
+ *  move the new objects into the repo that would produce an repo that has a broken
+ *  object reference. There is nothing that prohibits two parallel commits to the
+ *  same branch, and doing that could cause one of the commits to be lost in the
+ *  branch history. However, the repo as a whole will alwyas end up valid.
+ *
+ * NOTHING:
+ *  Operations that are purely read-only and can either succeed or
+ *  not as a whole do nothing to protect against parallelism. Typical examples
+ *  are checkouts or pulls from a remote client. If such an operation is started
+ *  nothing protects the repo from removing (by e.g. prune) objects from the repo
+ *  that will be necessary to complete the operation. However, such an issue will
+ *  be detected by the operation.
+ *
+ * Given the above the standard approach for locking during prune should be to take
+ * an exclusive lock during the entire operation. However, the initial scan of the
+ * reachable objects of a repo can take a very long time, and blocking any new
+ * commits during this is not a great idea. So, to avoid this the prune operation
+ * does two scans of the reachable commits. One with a shared lock and then again
+ * with an exclusive lock.
```

Comment:
Maybe worth explaining that the time spent with the exclusive lock held will be shorter than if it were a single scan, because we can safely skip scanning commits that were enumerated in the previous shared-lock scan.